File: kotlin/src/main/kotlin/com/attomus/otp/OTPURIParser.kt
package com.attomus.otp

import java.io.ByteArrayOutputStream
import java.nio.charset.CharacterCodingException
import java.math.BigInteger
import java.net.URI
import java.net.URISyntaxException
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.util.Date
import java.util.UUID

@Suppress("TooManyFunctions")
object OTPURIParser {
    @Suppress("ThrowsCount")
    fun parse(uri: String): OTPProvisioningResult {
        val components = parseComponents(uri)
        val type = parseType(components)
        val label = parseLabel(components.rawPath)
        val parameters = parseParameters(components.rawQuery)

        val encodedSecret = parameters["secret"]
        if (encodedSecret.isNullOrEmpty()) {
            throw OTPURIError.MissingSecret
        }

        val secret = try {
            Base32.decode(encodedSecret)
        } catch (_: Base32Error) {
            throw OTPURIError.InvalidSecret
        }

        try {
            OTPValidation.validateURISecret(secret)

            return OTPProvisioningResult(
                account = makeAccount(type, label, parameters),
                secret = secret
            )
        } catch (error: OTPURIError) {
            secret.zero()
            throw error
        }
    }

    @Suppress("ThrowsCount")
    private fun parseComponents(uri: String): URI {
        if (uri.length > 2048) {
            throw OTPURIError.UriTooLong
        }

        validatePercentEncoding(uri)

        val components = try {
            URI(uri)
        } catch (_: URISyntaxException) {
            throw OTPURIError.InvalidScheme
        } catch (_: IllegalArgumentException) {
            throw OTPURIError.InvalidScheme
        }

        if (!components.scheme.equals("otpauth", ignoreCase = true)) {
            throw OTPURIError.InvalidScheme
        }

        return components
    }

    private fun parseType(components: URI): OTPType {
        val rawType = components.host ?: throw OTPURIError.InvalidType
        return when {
            rawType.equals("totp", ignoreCase = true) -> OTPType.TOTP
            rawType.equals("hotp", ignoreCase = true) -> OTPType.HOTP
            else -> throw OTPURIError.InvalidType
        }
    }

    @Suppress("ThrowsCount")
    private fun parseLabel(rawPath: String?): LabelParts {
        val path = rawPath?.removePrefix("/") ?: ""
        if (path.isEmpty()) {
            throw OTPURIError.MissingLabel
        }

        val decoded = percentDecode(path)
        val parts = decoded.split(":", limit = 2)
        return if (parts.size == 2) {
            val issuer = parts[0].trim().ifEmpty { null }
            val accountName = parts[1].trim()
            if (accountName.isEmpty()) {
                throw OTPURIError.MissingLabel
            }
            LabelParts(issuer = issuer, accountName = accountName)
        } else {
            val accountName = decoded.trim()
            if (accountName.isEmpty()) {
                throw OTPURIError.MissingLabel
            }
            LabelParts(issuer = null, accountName = accountName)
        }
    }

    private fun parseParameters(rawQuery: String?): Map<String, String> {
        if (rawQuery.isNullOrEmpty()) {
            return emptyMap()
        }

        val parameters = linkedMapOf<String, String>()
        rawQuery.split("&")
            .asSequence()
            .filter { it.isNotEmpty() }
            .forEach { pair ->
                val equalsIndex = pair.indexOf('=')
                val rawName = if (equalsIndex >= 0) pair.substring(0, equalsIndex) else pair
                val rawValue = if (equalsIndex >= 0) pair.substring(equalsIndex + 1) else ""
                val name = percentDecode(rawName).lowercase()
                val value = percentDecode(rawValue)
                parameters.putIfAbsent(name, value)
            }

        return parameters
    }

    private fun makeAccount(
        type: OTPType,
        label: LabelParts,
        parameters: Map<String, String>
    ): OTPAccount {
        val algorithm = parseAlgorithm(parameters["algorithm"])
        val digits = parseDigits(parameters["digits"])
        val period = parsePeriod(parameters["period"], type)
        val issuer = resolveIssuer(label.issuer, parameters["issuer"])
        val counter = if (type == OTPType.HOTP) parseCounter(parameters["counter"]) else null

        return OTPAccount(
            id = UUID.randomUUID(),
            type = type,
            label = label.accountName,
            issuer = issuer,
            algorithm = algorithm,
            digits = digits,
            period = period,
            counter = counter,
            createdAt = Date()
        )
    }

    private fun parseAlgorithm(rawValue: String?): OTPAlgorithm {
        val value = rawValue?.trim().orEmpty()
        if (value.isEmpty()) {
            return OTPAlgorithm.SHA1
        }

        return when (value.lowercase()) {
            "sha1" -> OTPAlgorithm.SHA1
            "sha256" -> OTPAlgorithm.SHA256
            "sha512" -> OTPAlgorithm.SHA512
            else -> throw OTPURIError.InvalidAlgorithm
        }
    }

    private fun parseDigits(rawValue: String?): Int {
        val value = rawValue?.trim().orEmpty()
        if (value.isEmpty()) {
            return 6
        }

        val digits = value.toIntOrNull() ?: throw OTPURIError.InvalidDigits
        OTPValidation.validateURIDigits(digits)
        return digits
    }

    @Suppress("ReturnCount")
    private fun parsePeriod(rawValue: String?, type: OTPType): Int {
        return when {
            type == OTPType.HOTP -> 30
            rawValue.isNullOrBlank() -> 30
            else -> {
                val period = rawValue.trim().toIntOrNull() ?: throw OTPURIError.InvalidPeriod
                OTPValidation.validateURIPeriod(period)
                period
            }
        }
    }

    @Suppress("ThrowsCount")
    private fun parseCounter(rawValue: String?): Long {
        val value = rawValue?.trim().orEmpty()
        if (value.isEmpty()) {
            throw OTPURIError.MissingCounter
        }

        val counter = try {
            BigInteger(value)
        } catch (_: Exception) {
            throw OTPURIError.MissingCounter
        }

        if (counter.signum() < 0 || counter.bitLength() > 64) {
            throw OTPURIError.MissingCounter
        }

        return counter.toLong()
    }

    private fun resolveIssuer(labelIssuer: String?, parameterIssuer: String?): String? {
        val normalizedParameter = parameterIssuer?.trim().orEmpty().ifEmpty { null }
        if (labelIssuer != null && normalizedParameter != null && labelIssuer != normalizedParameter) {
            throw OTPURIError.IssuerMismatch
        }
        return normalizedParameter ?: labelIssuer
    }

    private fun validatePercentEncoding(text: String) {
        var index = 0
        while (index < text.length) {
            if (text[index] == '%') {
                if (index + 2 >= text.length) {
                    throw OTPURIError.MalformedPercentEncoding
                }
                val first = text[index + 1]
                val second = text[index + 2]
                if (!first.isHexDigit() || !second.isHexDigit()) {
                    throw OTPURIError.MalformedPercentEncoding
                }
                index += 3
                continue
            }
            index += 1
        }
    }

    private fun percentDecode(text: String): String {
        val output = StringBuilder()
        val bytes = ByteArrayOutputStream()

        fun flushBytes() {
            if (bytes.size() == 0) {
                return
            }

            val decoder = StandardCharsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)

            val decoded = try {
                decoder.decode(java.nio.ByteBuffer.wrap(bytes.toByteArray())).toString()
            } catch (_: CharacterCodingException) {
                throw OTPURIError.MalformedPercentEncoding
            }

            output.append(decoded)
            bytes.reset()
        }

        var index = 0
        while (index < text.length) {
            val character = text[index]
            if (character == '%') {
                if (index + 2 >= text.length) {
                    throw OTPURIError.MalformedPercentEncoding
                }

                val first = text[index + 1]
                val second = text[index + 2]
                if (!first.isHexDigit() || !second.isHexDigit()) {
                    throw OTPURIError.MalformedPercentEncoding
                }

                val byteValue = ((hexValue(first) shl 4) or hexValue(second)).toByte()
                bytes.write(byteValue.toInt())
                index += 3
                continue
            }

            flushBytes()
            output.append(character)
            index += 1
        }

        flushBytes()
        return output.toString()
    }

    private fun Int.isHexDigit(): Boolean = this in 0..15

    private fun Char.isHexDigit(): Boolean = hexValue(this) >= 0

    private fun hexValue(character: Char): Int {
        return when (character) {
            in '0'..'9' -> character - '0'
            in 'a'..'f' -> 10 + (character - 'a')
            in 'A'..'F' -> 10 + (character - 'A')
            else -> -1
        }
    }

    private fun ByteArray.zero() {
        fill(0)
    }

    private data class LabelParts(
        val issuer: String?,
        val accountName: String
    )
}

File: swift/Sources/AttomusOTP/OTPURIParser.swift
import Foundation

public func parseOTPURI(_ uri: String) throws -> OTPProvisioningResult {
    let components = try parseComponents(from: uri)
    let type = try parseType(from: components)
    let label = try parseLabel(from: components.percentEncodedPath)
    let parameters = parseParameters(from: components)
    guard let encodedSecret = parameters["secret"], !encodedSecret.isEmpty else {
        throw OTPURIError.missingSecret
    }

    let secret: Data
    do {
        secret = try Base32.decode(encodedSecret)
    } catch {
        throw OTPURIError.invalidSecret
    }

    try OTPValidation.validateURISecret(secret)

    if type == .hotp {
        _ = try parseCounter(parameters["counter"])
    }

    return try OTPProvisioningResult(
        account: makeAccount(type: type, label: label, parameters: parameters),
        secret: secret
    )
}

private func parseComponents(from uri: String) throws -> URLComponents {
    guard uri.count <= 2048 else {
        throw OTPURIError.uriTooLong
    }

    try validatePercentEncoding(in: uri)

    guard let components = URLComponents(string: uri) else {
        throw OTPURIError.invalidScheme
    }

    guard components.scheme?.caseInsensitiveCompare("otpauth") == .orderedSame else {
        throw OTPURIError.invalidScheme
    }

    return components
}

private func parseType(from components: URLComponents) throws -> OTPType {
    guard let rawType = components.host?.lowercased() else {
        throw OTPURIError.invalidType
    }

    switch rawType {
    case "totp":
        return .totp
    case "hotp":
        return .hotp
    default:
        throw OTPURIError.invalidType
    }
}

private func parseParameters(from components: URLComponents) -> [String: String] {
    Dictionary(
        components.queryItems?.map { ($0.name.lowercased(), $0.value ?? "") } ?? [],
        uniquingKeysWith: { first, _ in first }
    )
}

private func makeAccount(
    type: OTPType,
    label: (issuer: String?, accountName: String),
    parameters: [String: String]
) throws -> OTPAccount {
    let algorithm = try parseAlgorithm(parameters["algorithm"])
    let digits = try parseDigits(parameters["digits"])
    let period = try parsePeriod(parameters["period"], type: type)
    let issuer = try resolveIssuer(labelIssuer: label.issuer, parameterIssuer: parameters["issuer"])

    return OTPAccount(
        id: UUID(),
        type: type,
        label: label.accountName,
        issuer: issuer,
        algorithm: algorithm,
        digits: digits,
        period: period,
        counter: type == .hotp ? try parseCounter(parameters["counter"]) : nil,
        createdAt: Date()
    )
}

private func parseLabel(from percentEncodedPath: String) throws -> (issuer: String?, accountName: String) {
    let trimmedPath = percentEncodedPath.hasPrefix("/") ? String(percentEncodedPath.dropFirst()) : percentEncodedPath
    guard !trimmedPath.isEmpty else {
        throw OTPURIError.missingLabel
    }

    guard let decoded = trimmedPath.removingPercentEncoding else {
        throw OTPURIError.malformedPercentEncoding
    }

    let parts = decoded.split(separator: ":", maxSplits: 1, omittingEmptySubsequences: false).map(String.init)
    if parts.count == 2 {
        let issuer = parts[0].trimmingCharacters(in: .whitespaces)
        let accountName = parts[1].trimmingCharacters(in: .whitespaces)
        guard !accountName.isEmpty else {
            throw OTPURIError.missingLabel
        }
        return (issuer.isEmpty ? nil : issuer, accountName)
    }

    let accountName = decoded.trimmingCharacters(in: .whitespaces)
    guard !accountName.isEmpty else {
        throw OTPURIError.missingLabel
    }
    return (nil, accountName)
}

private func parseAlgorithm(_ rawValue: String?) throws -> OTPAlgorithm {
    guard let rawValue, !rawValue.isEmpty else {
        return .sha1
    }

    switch rawValue.lowercased() {
    case "sha1":
        return .sha1
    case "sha256":
        return .sha256
    case "sha512":
        return .sha512
    default:
        throw OTPURIError.invalidAlgorithm
    }
}

private func parseDigits(_ rawValue: String?) throws -> Int {
    guard let rawValue, !rawValue.isEmpty else {
        return 6
    }

    guard let digits = Int(rawValue) else {
        throw OTPURIError.invalidDigits
    }

    try OTPValidation.validateURIDigits(digits)
    return digits
}

private func parsePeriod(_ rawValue: String?, type: OTPType) throws -> Int {
    guard type == .totp else {
        return 30
    }

    guard let rawValue, !rawValue.isEmpty else {
        return 30
    }

    guard let period = Int(rawValue) else {
        throw OTPURIError.invalidPeriod
    }

    try OTPValidation.validateURIPeriod(period)
    return period
}

private func parseCounter(_ rawValue: String?) throws -> UInt64 {
    guard let rawValue, !rawValue.isEmpty else {
        throw OTPURIError.missingCounter
    }

    guard !rawValue.hasPrefix("-"),
          let counter = UInt64(rawValue),
          counter <= UInt64(Int64.max) else {
        throw OTPURIError.missingCounter
    }

    return counter
}

private func resolveIssuer(labelIssuer: String?, parameterIssuer: String?) throws -> String? {
    let decodedParameter = parameterIssuer?.trimmingCharacters(in: .whitespacesAndNewlines)
    let normalizedParameter = decodedParameter?.isEmpty == true ? nil : decodedParameter

    if let labelIssuer, let normalizedParameter, labelIssuer != normalizedParameter {
        throw OTPURIError.issuerMismatch
    }

    return normalizedParameter ?? labelIssuer
}

private func validatePercentEncoding(in text: String) throws {
    var index = text.startIndex

    while index < text.endIndex {
        if text[index] == "%" {
            let first = text.index(after: index)
            guard first < text.endIndex else {
                throw OTPURIError.malformedPercentEncoding
            }

            let second = text.index(after: first)
            guard second < text.endIndex else {
                throw OTPURIError.malformedPercentEncoding
            }

            let pair = text[first...second]
            guard pair.allSatisfy(\.isHexDigit) else {
                throw OTPURIError.malformedPercentEncoding
            }

            index = text.index(after: second)
            continue
        }

        index = text.index(after: index)
    }
}
